### PR TITLE
Dialogue: edit participant from canvas

### PR DESCRIPTION
**Pull request by retrofox**
This PR implements a new component used to edit the participants in a conversation structure.

https://user-images.githubusercontent.com/77539/106919456-ec6dd980-66e8-11eb-96a3-92f5f520cc12.mov

#### Changes proposed in this Pull Request:

* Dialogue: edit participant from canvas

#### Does this pull request change what data or activity we track or use?

NO

#### Testing instructions:

* Create a new conversation block. Check you can get it by `conversation`, `transcription`, `dialogue`, `dialog`, and `speaker` keywords.
* When you start a new component from scratch, there is not any participant defined in the conversation context, thus you need to start to type the participant name in some of the dialogue blocks.
* The new participant will be added once press `ENTER` key or when on focus outside
* When editing a new Dialogue participant, you should get the participant suggestions.
* To add a new participant, you need to clean the participant content before starting to type. Otherwise, it will edit the current participant and all sibling Dialogue blocks with the same participant.

D56506-code

#### Proposed changelog entry for your changes:

*

### CODE REVIEW

**Comment by sfougnier** [comment]
🎉 Overall this is great! Adding/selecting speakers and only showing 1 child block initially feels much more intuitive and easy to use. 

A few things that stand out:

- We've already discussed the positioning of the autocomplete, but want to flag that again here. When the autocomplete appears above the text entry rather than below, the placement feels unexpected based on existing autocomplete patterns.
- It also obstructs the view of the UI in the toolbar. If I hadn't seen your previous screen recordings, I wouldn't know I could edit the speaker label text styling. _(see screenshot below)_
Screen Shot 2021-02-02 at 11 38 23 AM

I wonder if we could make the speaker label text **bold** by default to distinguish it from the transcription text and move the label color settings to the sidebar so that it doesn't get lost.

File: projects/plugins/jetpack/extensions/blocks/conversation/edit.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/conversation/edit.js:45]: > * We've already discussed the positioning of the autocomplete, but want to flag that again here. When the autocomplete appears above the text entry rather than below, the placement feels unexpected based on existing autocomplete patterns.

Not sure what do you exactly mean with  _existing autocomplete patterns_. I think all autcompleters in core behave in the same way. It's a popover with a [default `top right` position](https://github.com/WordPress/gutenberg/blob/d8d208927e964bf4f928a9e6b378902ed5fc421e/packages/components/src/autocomplete/index.js#L244).
**Is it a blocker?**
Btw, [I created a PR](https://github.com/WordPress/gutenberg/pull/28611) that allows extending the Popover properties, in the same that other components do, like the Dropdown component. It won't be an immediate solution but if it's merged in the core, it will be updated in the near future.

> It also obstructs the view of the UI in the toolbar.

The autocomplete always overlaps over the canvas area. It happens for all of them. Typing `/` in a paragraph block, for instance.



Even more, if you press `TAB` it won't be possible to hide the blocks selector via keyboard. In that case, you need to click on another canvas area.

I've updated this PR forcing to hide the autocomplete when the component loses the focus, via the `TAB` key for instance. The slightly bad news is the `onFocusOutside` event isn't available in the current Gutenberg version for Jetpack. It will work in other sites with newer core versions.

> If I hadn't seen your previous screen recordings, I wouldn't know I could edit the speaker label text styling. (see screenshot below)

I think it's a matter to hide the autocomplete. Just selects the participant and the popover will disappear or pressing the `ESC` key. The previous comment above will avoid this situation, too.

> I wonder if we could make the speaker label text bold by default to distinguish it from the transcription text and move the label color settings to the sidebar so that it doesn't get lost.

Trying....

**Comment by retrofox** [comment]
> * We've already discussed the positioning of the autocomplete, but want to flag that again here. When the autocomplete appears above the text entry rather than below, the placement feels unexpected based on existing autocomplete patterns.

Not sure what do you exactly mean with  _existing autocomplete patterns_. I think all autcompleters in core behave in the same way. It's a popover with a [default `top right` position](https://github.com/WordPress/gutenberg/blob/d8d208927e964bf4f928a9e6b378902ed5fc421e/packages/components/src/autocomplete/index.js#L244).
**Is it a blocker?**
Btw, [I created a PR](https://github.com/WordPress/gutenberg/pull/28611) that allows extending the Popover properties, in the same that other components do, like the Dropdown component. It won't be an immediate solution but if it's merged in the core, it will be updated in the near future.

> It also obstructs the view of the UI in the toolbar.

The autocomplete always overlaps over the canvas area. It happens for all of them. Typing `/` in a paragraph block, for instance.



Even more, if you press `TAB` it won't be possible to hide the blocks selector via keyboard. In that case, you need to click on another canvas area.

I've updated this PR forcing to hide the autocomplete when the component loses the focus, via the `TAB` key for instance. The slightly bad news is the `onFocusOutside` event isn't available in the current Gutenberg version for Jetpack. It will work in other sites with newer core versions.

> If I hadn't seen your previous screen recordings, I wouldn't know I could edit the speaker label text styling. (see screenshot below)

I think it's a matter to hide the autocomplete. Just selects the participant and the popover will disappear or pressing the `ESC` key. The previous comment above will avoid this situation, too.

> I wonder if we could make the speaker label text bold by default to distinguish it from the transcription text and move the label color settings to the sidebar so that it doesn't get lost.

Trying....

File: projects/plugins/jetpack/extensions/blocks/conversation/edit.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/conversation/edit.js]: https://github.com/Automattic/jetpack/pull/18646/commits/7de2e0b9f9775e906b5409f795d532378a29ddd6

**Comment by sfougnier** [comment]
> Not sure what do you exactly mean with existing autocomplete patterns. I think all autcompleters in core behave in the same way. It's a popover with a default top right position.

I mean more common autocomplete patterns across the web rather than specific to WordPress, thinking of these users who are less familiar with Gutenberg.

> Is it a blocker?

No, not a blocker. Apologies if it came off that way. I know we're working on trying to get this into core, and I still think this behavior is more intuitive than what we had previously. I was using the observation to lead to suggesting moving color editing to the sidebar.

> I think it's a matter to hide the autocomplete. Just selects the participant and the popover will disappear or pressing the ESC key. The previous comment above will avoid this situation, too.

Ok cool, let's try it!

**Comment by retrofox** [comment]
> > Not sure what do you exactly mean with existing autocomplete patterns. I think all autcompleters in core behave in the same way. It's a popover with a default top right position.
> 
> I mean more common autocomplete patterns across the web rather than specific to WordPress, thinking of these users who are less familiar with Gutenberg.

👍 


> > Is it a blocker?
> 
> No, not a blocker. Apologies if it came off that way. I know we're working on trying to get this into core, and I still think this behavior is more intuitive than what we had previously. I was using the observation to lead to suggesting moving color editing to the sidebar.

Yeah, nice. Surely it's something that we can keep improving but asking in case we should change the approach.  



> > I think it's a matter to hide the autocomplete. Just selects the participant and the popover will disappear or pressing the ESC key. The previous comment above will avoid this situation, too.

> Ok cool, let's try it!

Sorry, maybe I wasn't clear. My point is in some way it's the natural behavior of the autocomplete. However, we are trying to control it a little bit more. For instance, hiding it in the `edit` mode.

**Comment by retrofox** [comment]
Updated blocks icons
image

**Comment by mreishus** [comment]
I haven't been following this closely, but I just tried the new conversation block.  Here's two things I noticed:

`id: 2021-02-02-mmr-1`
New Post -> Click `+` -> Search for Conversation -> Click Conversation to add the block -> Cursor is in speaker -> Type "Alice" and press enter.  
  * Expect to see: Cursor in "Write Dialog..." where I can type and provide Alice's dialog. 
  * Actually see: I have to also press TAB before I can start typing dialog. (So "Alice", Enter, Tab, Type)

`id: 2021-02-02-mmr-2`
Create a dialog block with a speaker but without any text.

Click anywhere in "Write Dialog..."
  * Expect to see: Cursor blinking somewhere in "Write Dialog...", indicating that I can type.
  * Actually see: No cursor.  If I type, I fill it out, which is good, but I would expect to see a cursor.

Note that if I click, type "a", then backspace, I see the placeholder text and the cursor.


`id: 2021-02-02-mmr-3`
This one was tricky to explain, so I made a video. 

https://user-images.githubusercontent.com/937354/106675243-6764b200-657a-11eb-8129-8e9363540eff.mp4
Unfortunately the drop down does not appear in the video, so I made a screenshot:

**Comment by sfougnier** [comment]
The latest changes are looking good! I love the keyboard shortcuts to tab between the elements and hit enter to autocomplete.

Regarding deleting speaker names, I don't think it's a big issue for launch:
- When I'm working in the canvas, it seems most intuitive to me that deleting all instances of a name won't cause it to show in the autocomplete. There's now less cognitive load in adding, deleting, selecting. I'm just removing things as I would in a text editor.
- In the sidebar, I do see the issue with still having deleted names show in that dropdown. I think it's something to think through further, but I don't think the current state should be a blocker or a priority to work through for launch.

**Comment by creativecoder** [comment]
FYI @retrofox, I merged the changes from https://github.com/Automattic/jetpack/pull/18598 into this PR to resolve the conflicts.

**Comment by retrofox** [comment]
> `id: 2021-02-02-mmr-1`
> New Post -> Click `+` -> Search for Conversation -> Click Conversation to add the block -> Cursor is in speaker -> Type "Alice" and press enter.
> 
> * Expect to see: Cursor in "Write Dialog..." where I can type and provide Alice's dialog.
> * Actually see: I have to also press TAB before I can start typing dialog. (So "Alice", Enter, Tab, Type)

 removed because it's a little bit tricky to deal with, or at least it's different as I use to do. In short, keyboard events are silent in RichText. However, we can try to achieve this using `onSplit()` and/or `onReplace()`.

> 
> `id: 2021-02-02-mmr-2`
> Create a dialog block with a speaker but without any text.
> 
> Click anywhere in "Write Dialog..."
> 
> * Expect to see: Cursor blinking somewhere in "Write Dialog...", indicating that I can type.
> * Actually see: No cursor.  If I type, I fill it out, which is good, but I would expect to see a cursor.

I noted this too. We can remove the placeholder once the curser is right there. However, I've seen no curser even when no placeholder in the input.

>   Note that if I click, type "a", then backspace, I see the placeholder text and the cursor.

This is something that may happen also in other blocks. I guess it's core.


> This one was tricky to explain, so I made a video.
> 
> https://user-images.githubusercontent.com/937354/106675243-6764b200-657a-11eb-8129-8e9363540eff.mp4

Video is awesome as usual. Let me try to keep polishing the behavior. The trickier part is onFocusOutside is not available currently on the Jetpack site.

**Comment by retrofox** [comment]
Second round :-D

> `id: 2021-02-02-mmr-3`
> This one was tricky to explain, so I made a video.

`00:37`: Pressing `ENTER` of `TAB`, under those conditions, is the same. It adds the new speaker to the conversation. You can visually check that because:

* `CHA` doesn't have transparence in the canvas
* `CHA` is defined as the speaker name, in the block sidebar.

Screen Shot 2021-02-03 at 9 51 59 AM

`00:50`. It's a very nice feedback:



When you _delete_ it, actually you are editing the current participant name on the fly. It keeps editing the name _**until**_ the input is empty. From that, you start to _**add**_ a new speaker.
Maybe, what we should do is apply the chnages either when the user presses `ENTER` or `TAB` key.

> 

If `C` is right there, the speaker does exist.

File: projects/plugins/jetpack/extensions/blocks/conversation/participants.json
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/conversation/participants.json:7]: yeah, sounds good to me. I left the plain participant (label) because maybe it could be useful to print a participant list in areas without formatting. In case we don't wanna that, we can remove it. 👍

File: projects/plugins/jetpack/extensions/blocks/conversation/edit.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/conversation/edit.js:46]: sounds good 👍

**Comment by sfougnier** [comment]
After testing the block performing a transcription using some of the keyboard shortcuts:
- Hitting `Enter` after typing a name allows me to add it and move to the next line.
- I can also `Tab` from the speaker text to the transcript text.
- When I have a few letters (He) typed from a name already entered previously (Henry), I can hit `Enter` in the autocomplete dropdown and that will populate Henry in the speaker name. At this point, can the cursor automatically move to the end of the name so that I could hit `Enter` and keep typing my transcript?

https://user-images.githubusercontent.com/67021691/106822515-491bb680-664d-11eb-9793-df33d7d0d2b9.mov


- Similarly, if I'm typing my transcription text and hit `Enter` could that just give me a new line break, and then hitting the key twice `Enter Enter` adds the next Transcript block below?

**Review by creativecoder — COMMENTED** [review]
Here are 3 usability issues I noticed while testing out the latest version

I notice that editing and deleting participants is available in the parent conversation block, but has no effect. Maybe we should remove that interface, at least for now?

edit-participants-from-conversation

---

Editing a participant, and then clicking outside the block causes the edit to reset. (This is with WordPress 5.6, but without the Gutenberg plugin active, so I'm guessing it's because `useFocusOutside` isn't available 

edit-participant-click-outside-block
 
---

I can't seem to paste dialogue content

dialogue-block-cant-paste

**Comment by sfougnier** [comment]
>Looking at the standard set of formatting controls in a paragraph, I think at least these should be available for participant name:
bold
italic
text color
link
underline

I agree that the current controls for the name seem a bit limited, but I also worry we're trying to add too many bells and whistles. 

Showing the name as 'regular' weight by default doesn't present a clear visual hierarchy between the different types of inputs. The speaker name should be **bold** by default to make it clearly distinguished from the paragraph text and either:
- Show bold toggled on in the toolbar, as well as the full amount of text editing options _or_
- Only allow for color to be changed, perhaps in the sidebar as reflected in the design.

**Comment by retrofox** [comment]
> After testing the block performing a transcription using some of the keyboard shortcuts:
> 
> * Hitting `Enter` after typing a name allows me to add it and move to the next line.

I think it's ok right?

> * I can also `Tab` from the speaker text to the transcript text.

it is ok, too.

> * When I have a few letters (He) typed from a name already entered previously (Henry), I can hit `Enter` in the autocomplete dropdown and that will populate Henry in the speaker name. At this point, can the cursor automatically move to the end of the name so that I could hit `Enter` and keep typing my transcript?

Probably. 

> * Similarly, if I'm typing my transcription text and hit `Enter` could that just give me a new line break, and then hitting the key twice `Enter Enter` adds the next Transcript block below?

Yes, I think so. But it will require analyzing the transcript content even more in detail. But I think it's doable.

Thanks, Saul. Very nice and valid ideas :+1: Let's try to achieve it. Honestly and on another hand, I'm a little bit concerned about keep adding changes here vs try to ship as it is and then iterate and polish later, in follow-up tasks.
I'd say that unless the suggestions are 100% percent needed, and given the current circumstances, let's try to close this one asap and keep working on it after landing.

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js:234]: Let me update it.

**Comment by retrofox** [comment]
> Here are 3 usability issues I noticed while testing out the latest version
> 
> I notice that editing and deleting participants is available in the parent conversation block, but has no effect. Maybe we should remove that interface, at least for now?

Yes, totally. We already have [a PR ](https://github.com/Automattic/jetpack/pull/18674) standing in the line for this.

> Editing a participant, and then clicking outside the block causes the edit to reset. (This is with WordPress 5.6, but without the Gutenberg plugin active, so I'm guessing it's because `useFocusOutside` isn't available

Correct. I tried to mitigate this lack of event handling by adding a custom behavior when onFocus() occurs in the transcript content component. It's not ideal but will help the UX.
Another thing that we could do is adding a component just next to the participant input, something that helps the user to see
* what's the current state of the input content: adding a new, selecting, or editing participant.
* it could be actionable (a button for instance) so the user can press it to apply the changes.

> I can't seem to paste dialogue content

Something to deal with in another PR, IMO. 

Thanks, Grant for your feedback. 👍

**Comment by retrofox** [comment]
> > Looking at the standard set of formatting controls in a paragraph, I think at least these should be available for participant name:
> > bold
> > italic
> > text color
> > link
> > underline
> 
> I agree that the current controls for the name seem a bit limited, but I also worry we're trying to add too many bells and whistles.

IMHO, we can provide the format options. In the end, will be the user who gets the final word.

> Showing the name as 'regular' weight by default doesn't present a clear visual hierarchy between the different types of inputs. The speaker name should be **bold** by default to make it clearly distinguished from the paragraph text and either:

Agree. Let's try to get it on this PR.


> * Show bold toggled on in the toolbar, as well as the full amount of text editing options _or_
> * Only allow for color to be changed, perhaps in the sidebar as reflected in the design.

Adding colors handling in the sidebar would be nice, following the designs. Just in case, it isn't part of the purpose of this PR. It's growing too much IMO.
Also, and again, considering the next dates, I'm not sure it's doable for v1. Even more, it's something that we agreed it should be addressed for v2 when we started to work on this feature.

It doesn't mean that we can try it for v1 😅 . Thanks, Saul.

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js:234]: > Editing a participant, and then clicking outside the block causes the edit to reset. (This is with WordPress 5.6, but without the Gutenberg plugin active, so I'm guessing it's because useFocusOutside isn't available

Forgot to mention that theoretically, Jetpack sites will support `onFocusOutside` soon. I say `theoretically` because the hook is still experimental :-/ I
I've added a fallback trivial function in case it's not defined in the WordPress package.

**Comment by retrofox** [comment]
Provable going to be replaced by https://github.com/Automattic/jetpack/pull/18711

**Review by pablinos — COMMENTED** [review]
Looks good. It's a fairly intuitive experience. I think it's always going to take a little bit of getting used to, but I can see this working. 

Things I've seen are:

* With a dark mod version of Spearhead the current time is showing in white on the toolbar
Screenshot 2021-02-08 at 18 08 35
* A weird thing, from time to time, where the cursor isn't rendered. I think it's something to do with either Firefox or an add on I have installed. I'm not seeing the same thing in Chrome.
* If you have a participant that is a substring of others, then the autocomplete disappears at the point that you have typed the full name of the participant that is a substring. It's not really a big deal as you can carry on typing or delete a character, but it might be nice to keep it open.
* Clicking to the right of the participant name selects the block, but it would be nice to have that position the cursor at the end of name ready to edit it.

I'm coming to the code a bit late, so it's tricky to work out everything that's going on. I've made a few comments, but it's obviously gone through a lot of iterations, so it might be best to get this merged and make updates in further PRs.

File: projects/plugins/jetpack/extensions/blocks/conversation/components/participants-controls.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/conversation/components/participants-controls.js:40]: it's in the conversation context. Where your there?

File: projects/plugins/jetpack/extensions/blocks/dialogue/edit.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/edit.js:67]: do you have a suggestion value?

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js:254]: let me take a look. Could you help me to reproduce the issue? I'm forcing a rendering to update the autocomplete, and it should be done only when the `participants` object changes.

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js]: removed here https://github.com/Automattic/jetpack/pull/18646/commits/2cdcf2a771a61bdb400ffa89a5c6eeef7be312bf

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js
Comment (pablinos) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js]: Yeah, it's a bit sporadic. I see you've removed the reducer, so I'll give it another try and see if I can get a reliable test case.

Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js]: removed here https://github.com/Automattic/jetpack/pull/18646/commits/2cdcf2a771a61bdb400ffa89a5c6eeef7be312bf

I had to add it because the speaker's list did not update once a new one was added. I've tested again, removing this technique and it seems to work. properly. 👍

File: projects/plugins/jetpack/extensions/blocks/dialogue/edit.js
Comment (pablinos) [reply projects/plugins/jetpack/extensions/blocks/dialogue/edit.js:67]: I'm not sure, 250 maybe? I'm not sure it matters if the other blocks updating is a little bit laggy, but it's probably something to try and see if it feels ok.

File: projects/plugins/jetpack/extensions/blocks/conversation/components/participants-controls.js
Comment (pablinos) [reply projects/plugins/jetpack/extensions/blocks/conversation/components/participants-controls.js:40]: Ah, thanks!

**Comment by allilevine** [comment]
Just double-checking, should there be a way to change the participant style? Or was that removed because we're going to add more options in the sidebar?
speakerstyles

**Comment by retrofox** [comment]
> Looks good. It's a fairly intuitive experience. I think it's always going to take a little bit of getting used to, but I can see this working.
> 
> Things I've seen are:
> 
> * With a dark mod version of Spearhead the current time is showing in white on the toolbar
> 
> Screenshot 2021-02-08 at 18 08 35

This is something not introduced by these changes so I suggest creating a new issue and try to tackle it in a follow-up PR. I'll keep an open eye.
 
> * A weird thing, from time to time, where the cursor isn't rendered. I think it's something to do with either Firefox or an add on I have installed. I'm not seeing the same thing in Chrome.

I've seen the same behavior even for paragraph blocks. Anyway, I'll start to test deeply in FF too. I've removed the re-rendering process so maybe not it works better.

> * If you have a participant that is a substring of others, then the autocomplete disappears at the point that you have typed the full name of the participant that is a substring. It's not really a big deal as you can carry on typing or delete a character, but it might be nice to keep it open.

This substring is split up by spaces?

> * Clicking to the right of the participant name selects the block, but it would be nice to have that position the cursor at the end of name ready to edit it.

I'm not sure if it has much sense, honestly. There are many reasons why I think it doesn't. It could work in the following disposition, for instance.

image

But if we add the timestamp, if changes the disposition.

image

Now, clicking to the right should make focus on the timestamp label, idk. A similar thing could happen when changing the block styles to cols.

> I'm coming to the code a bit late, so it's tricky to work out everything that's going on. I've made a few comments, but it's obviously gone through a lot of iterations, so it might be best to get this merged and make updates in further PRs.

:+1: Thanks for your review, Paul.

**Comment by retrofox** [comment]
> Just double-checking, should there be a way to change the participant style? Or was that removed because we're going to add more options in the sidebar?
> speakerstyles

we decided to remove the styles for now. Previously, it was possible to edit the styles through some of the toolbar buttons. Next, it was possible to use the RichText component. Now, it isn't possible and the speaker label has `<strong/>` tyle as default.

**Review by sfougnier — APPROVED** [review]
Looks good.

File: projects/plugins/jetpack/extensions/blocks/dialogue/edit.js
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/edit.js:67]: > Looks good. It's a fairly intuitive experience. I think it's always going to take a little bit of getting used to, but I can see this working.
> 
> Things I've seen are:
> 
> * With a dark mod version of Spearhead the current time is showing in white on the toolbar
> 
> Screenshot 2021-02-08 at 18 08 35

I think it should be addressed here https://github.com/Automattic/jetpack/pull/18756

Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/edit.js]: I've increased to 250 here https://github.com/Automattic/jetpack/pull/18646/commits/39e2db9849ecb6dc0f0d7f504bb84a2fe358cf26

**Comment by retrofox** [comment]
> Looks good. It's a fairly intuitive experience. I think it's always going to take a little bit of getting used to, but I can see this working.
> 
> Things I've seen are:
> 
> * With a dark mod version of Spearhead the current time is showing in white on the toolbar
> 
> Screenshot 2021-02-08 at 18 08 35

I think it should be addressed here https://github.com/Automattic/jetpack/pull/18756

**Comment by pablinos** [comment]
> I've seen the same behavior even for paragraph blocks. Anyway, I'll start to test deeply in FF too. I've removed the re-rendering process so maybe not it works better.

Yeah, I wouldn't be too concerned about it. The behaviour is inconsistent, and I suspect it's not specific to this block too.

> This substring is split up by spaces?

It happens the same with and without spaces. Say Paul, Paula, Pauline. Type Paul as the name and you will lose the autocomplete suggestions. It's a very minor problem, so we can probably address it (if we want to) in a follow-up.

> Now, clicking to the right should make focus on the timestamp label, idk. A similar thing could happen when changing the block styles to cols.

Good point, I didn't think about that! It just felt a bit tricky to click and focus at the end of the name, but given this, I think it's fine.

> I think it should be addressed here #18756

Thanks for that I'll head over there and give it a review.

**Comment by retrofox** [comment]
> > I've seen the same behavior even for paragraph blocks. Anyway, I'll start to test deeply in FF too. I've removed the re-rendering process so maybe not it works better.
> 
> Yeah, I wouldn't be too concerned about it. The behaviour is inconsistent, and I suspect it's not specific to this block too.

👍 

> 
> > This substring is split up by spaces?
> 
> It happens the same with and without spaces. Say Paul, Paula, Pauline. Type Paul as the name and you will lose the autocomplete suggestions. It's a very minor problem, so we can probably address it (if we want to) in a follow-up.

Got it.

https://user-images.githubusercontent.com/77539/107294614-7e107a80-6a4c-11eb-9f6c-d97d3a3caf0e.mov

**Comment by pablinos** [comment]
> Got it.

That's it, but is that likely to be more a general autocomplete component issue?

**Comment by pablinos** [comment]
I've noticed that removing a participant, doesn't completely remove it. Is that intentional? If it's in use then the speaker name is greyed out slightly. Should that be reset to the placeholder? I'm not sure as it might not be obvious enough that they're all related.

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js
Comment (pablinos) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js]: It's difficult to tell for sure, but it seems like I can't reproduce the focus problems I was seeing. I think you're right that this might be a more general problem with the editor, but I think for the context of this block we can consider it solved!

**Review by pablinos — APPROVED** [review]
I think this is good to go. There's more to be done, but we can make improvements in other PRs - as you're already doing! 😄

**Comment by retrofox** [comment]
> > Got it.
> 
> That's it, but is that likely to be more a general autocomplete component issue?

No really, because 

> > Got it.
> 
> That's it, but is that likely to be more a general autocomplete component issue?

No really. We were showing the autocomplete only when adding a new participant. I've added some changes to show it also when the speaker is going to be selected. https://github.com/Automattic/jetpack/pull/18646/commits/820d51d9fb56fa9fd490c14409da9592eeebd4d7

**Review by pablinos — APPROVED** [review]
Looks good! The autocomplete is still available when working with substring participants now.

**Review by jeherve — APPROVED** [review]
It seems to work well. I found 2 issues that I think would be blockers for launch, but maybe we can address them in follow-up PRs? I'll approve this PR and let you decide. 

*******

I miss the option to switch participants at times:

https://user-images.githubusercontent.com/426388/107373955-60442380-6ae7-11eb-97be-ba2256ddf482.mov

*********

If I publish a post with a conversation block, and then come back to edit it, when I hit enter it won't pre-fill the participant name, and I can't edit it later on:


https://user-images.githubusercontent.com/426388/107374224-a8634600-6ae7-11eb-8009-13b515fcab05.mov

**Comment by retrofox** [comment]
> It seems to work well. I found 2 issues that I think would be blockers for launch, but maybe we can address them in follow-up PRs? I'll approve this PR and let you decide.

> I miss the option to switch participants at times:

I'm not completely sure whether that's an issue. Basically, it filters by the speakers according to the caret position. In your video, the filtering string is `j`, getting two speakers `Joe` and `Jim`. But sometimes the filtering string is `ji`, in which case it returns only the `Jim` speaker label.

> If I publish a post with a conversation block, and then come back to edit it, when I hit enter it won't pre-fill the participant name, and I can't edit it later on:

Similar to the previous one. Let us dig on that. Basically, you need to start to type to get the popover. You can use and `space` to get all speakers. If you type `j`, it should show `Joe` and `Jim`.

**Comment by retrofox** [comment]
Thanks for helping with the review, folks.

**Comment by retrofox** [comment]
r220730-wpcom

**Comment by jeherve** [comment]
Cherry-picked to `jetpack/branch-9.4` in b111a3696a776c08918feb401abbadc448461d25

**Comment by jeherve** [comment]
> I'm not completely sure whether that's an issue. Basically, it filters by the speakers according to the caret position. In your video, the filtering string is j, getting two speakers Joe and Jim. But sometimes the filtering string is ji, in which case it returns only the Jim speaker label.

Ah, that's interesting. I didn't expect this to work that way, but now that you mention it I suppose it makes sense.

**Pull request merged**
